### PR TITLE
Implement abstract GasTree algorithm

### LINKED ISSUE #1072: Introduce abstract `GasTree` alrorithm implementation in `gear_common`
### File Location(s)

_No response_

### Proposal

Similarly to `pallet_gear_messenger` and it's `QueueImpl` type, we should implement gas tree algorithm in some trait (similarly to `common::storage::complex::queue::Queue`) and prepare `pallet_gas` to be an associated type for `pallet_gear::Config`. This opens a door to implementing gas logic in `gtest` and `gear_test`.

Must be resolved only after #1069.

### PR DESCRIPTION
**Pull request by gshep**
Resolves #1072 

- Introduce GasTree and GasAllowance traits. Add GasProvider trait exporting GasTree/GasAllowance
- Implement GasTree in common crate
- Remove dependency of pallet_gear::Config from pallet_gas
- Move property tests to common::gas_provider module

@gear-tech/dev

### CODE REVIEW

File: pallets/usage/src/tests.rs [hunk 1]
```diff
@@ -80,11 +81,8 @@ where
             ),
             blk_num,
         );
-        let _ = <Test as pallet_gear::Config>::GasHandler::create(
-            user_id.into_origin(),
-            msg_id.into_origin(),
-            gas_limit,
-        );
+        let _ =
+            GasHandlerOf::<Test>::create(user_id.into_origin(), msg_id.into_origin(), gas_limit);
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/usage/src/tests.rs:85]: for whole `GasTree` api

File: common/src/gas_provider/error.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    ///
+    /// If node is a parent to some other node it must have at least
+    /// one child, otherwise it's id can't be used as a parent for
+    /// local nodes in the tree.
+    fn parent_has_no_children() -> Self;
+
+    /// Value node doesn't exist for a key
+    fn node_not_found() -> Self;
+
+    /// Procedure can't be called on consumed node
```
Comment (breathx) [inline common/src/gas_provider/error.rs:45]:
```suggestion
    /// Procedure can't be called on consumed node.
```

Dots at the end of each doc comment.

File: common/src/gas_provider/mod.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+type GasBalanceKey<Balance, Key> = Option<(Balance, Key)>;
+type ConsumeOutput<Imbalance, External> = Option<(Imbalance, External)>;
+
+pub trait Allowance {
+    type Balance;
+
+    fn get() -> Self::Balance;
+    fn update(gas: Self::Balance);
+    fn decrease(gas: Self::Balance);
+}
```
Comment (breathx) [inline common/src/gas_provider/mod.rs:158]: This stuff already exists in storage primitives: check `complicated::Limiter`

File: common/src/gas_provider/node.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
+// This file is part of Gear.
```
Comment (breathx) [inline common/src/gas_provider/node.rs:1]: please, rename existing structures to `GasNode` and `GasNodeType`

File: pallets/gas/src/lib.rs [hunk 5]
```diff
@@ -226,565 +54,131 @@ pub mod pallet {
...
-            let second = self.0 - first;
+    #[pallet::storage]
+    pub type ValueTreeNodes<T> = StorageMap<_, Identity, H256, ValueNode>;
 
-            mem::forget(self);
-            (Self(first, PhantomData), Self(second, PhantomData))
-        }
+    common::wrap_storage_map!(
+        storage: ValueTreeNodes,
+        name: ValueTreeNodesWrap,
```
Comment (breathx) [inline pallets/gas/src/lib.rs:120]: `GasNodes`
`GasNodesWrap`

File: pallets/gas/src/lib.rs [hunk 6]
```diff
@@ -226,565 +54,131 @@ pub mod pallet {
...
 
-        fn drop_zero(self) -> Result<(), Self> {
-            if self.0.is_zero() {
-                Ok(())
-            } else {
-                Err(self)
-            }
-        }
+    pub type Key = H256;
+    pub type ExternalOrigin = H256;
```
Comment (breathx) [inline pallets/gas/src/lib.rs:112]:
```suggestion
    pub type Key = MessageId;
    pub type ExternalOrigin = Config::AccountId;
```

**Review by breathx — CHANGES_REQUESTED** [review]
Resolve given comments and provide docs for pallet:
be like this
https://github.com/gear-tech/gear/blob/96089c24a6180021dbdd1f4badb2efae8bd21d55/pallets/gear-messenger/src/lib.rs#L19
and this
https://github.com/gear-tech/gear/blob/96089c24a6180021dbdd1f4badb2efae8bd21d55/pallets/gear-messenger/src/lib.rs#L335

and so on

Great job at all! 👍 🚀

File: common/src/gas_provider/negative_imbalance.rs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+        self
+    }
+
+    fn subsume(&mut self, other: Self) {
+        self.0 = self.0.saturating_add(other.0);
+        mem::forget(other);
+    }
+
+    #[allow(clippy::comparison_chain)]
```
[the remark this answers is not in the document]
Comment (gshep) [reply common/src/gas_provider/negative_imbalance.rs:91]: Good remark, thanks! I just didn't dive too deep into the code when moving it around. Edited

**Review by breathx — APPROVED** [review]
Delicious ❤️

**Review by ekovalev — APPROVED** [review]

**Pull request merged**